message/validation/consensus_validation.go

```diff
@@ -228,52 +290,72 @@ func (mv *messageValidator) validateQBFTMessageByDutyLogic(
 	msgSlot := phase0.Slot(consensusMessage.Height)
 
 	for _, signer := range signedSSVMessage.GetOperatorIDs() {
-		signerState := state.GetSignerState(signer)
-		if signerState == nil {
-			signerState = state.CreateSignerState(signer)
-		}
-		if msgSlot > signerState.Slot {
-			newEpoch := mv.netCfg.Beacon.EstimatedEpochAtSlot(msgSlot) > mv.netCfg.Beacon.EstimatedEpochAtSlot(signerState.Slot)
-			signerState.ResetSlot(msgSlot, consensusMessage.Round, newEpoch)
-		} else if msgSlot == signerState.Slot && consensusMessage.Round > signerState.Round {
+		var signerState *SignerState
+
+		stateBySlot := consensusState.GetOrCreate(signer)
+		signerStateInterface, ok := stateBySlot.Get(msgSlot)
+		if !ok {
+			signerState = NewSignerState()
 			signerState.ResetRound(consensusMessage.Round)
-		}
 
-		if len(signedSSVMessage.FullData) != 0 && signerState.ProposalData == nil {
-			signerState.ProposalData = signedSSVMessage.FullData
+			stateBySlot.Put(msgSlot, signerState)
+			mv.pruneOldSlots(stateBySlot, msgSlot)
+		} else {
+			signerState = signerStateInterface.(*SignerState)
+			if consensusMessage.Round > signerState.Round {
+				signerState.ResetRound(consensusMessage.Round)
+			}
 		}
 
-		signerCount := len(signedSSVMessage.GetOperatorIDs())
-		if signerCount > 1 {
-			signerState.SeenDecidedLengths[signerCount] = struct{}{}
+		if err := mv.processSignerState(signedSSVMessage, consensusMessage, signerState); err != nil {
+			return err
 		}
-
-		signerState.MessageCounts.RecordConsensusMessage(signedSSVMessage, consensusMessage)
 	}
+
+	return nil
 }
 
-func (mv *messageValidator) validNumberOfCommitteeDutiesPerEpoch(
-	signedSSVMessage *spectypes.SignedSSVMessage,
-	validatorIndices []phase0.ValidatorIndex,
-	signerState *SignerState,
-	msgSlot phase0.Slot,
-) error {
-	newDutyInSameEpoch := false
-	if msgSlot > signerState.Slot && mv.netCfg.Beacon.EstimatedEpochAtSlot(msgSlot) == mv.netCfg.Beacon.EstimatedEpochAtSlot(signerState.Slot) {
-		newDutyInSameEpoch = true
+func (mv *messageValidator) processSignerState(signedSSVMessage *spectypes.SignedSSVMessage, consensusMessage *specqbft.Message, signerState *SignerState) error {
+	if len(signedSSVMessage.FullData) != 0 && consensusMessage.MsgType == specqbft.ProposalMsgType {
+		signerState.ProposalData = signedSSVMessage.FullData
 	}
 
-	if err := mv.validateDutyCount(validatorIndices, signerState, signedSSVMessage.SSVMessage.GetID(), newDutyInSameEpoch); err != nil {
-		return err
+	signerCount := len(signedSSVMessage.GetOperatorIDs())
+	if signerCount > 1 {
+		encodedOperators, err := encodeOperators(signedSSVMessage.GetOperatorIDs())
+		if err != nil {
+			// encodeOperators must never re
+			return ErrEncodeOperators
+		}
+
+		signerState.SeenSigners[string(encodedOperators)] = struct{}{}
 	}
 
+	signerState.MessageCounts.RecordConsensusMessage(signedSSVMessage, consensusMessage)
 	return nil
 }
 
+func (mv *messageValidator) pruneOldSlots(stateBySlot *treemap.Map, lastSlot phase0.Slot) {
```

Comment:
what is the reason for not just using a Go map here?

i prefer to not add a dependency if we dont have to

Reply:
@MatusKysel yes, but why do we even need this package at all? why are Go maps not enough? i strongly prefer to not add such big dependencies unless they're really worth the risks and other trade-offs.

For example the `ttlcache` package has a clear benefit where we use it (we TTL to reduce memory usage), and its a very small dependency (less risks) and very easy to reason about what its doing without having to go to their docs.

Reply:
if we use treemap in order to be able to tell the `Min` for pruning, i think instead we can just create our own struct which holds a standard slot map with a `minimumValue` field and `Put(slot, state)/Get(slot)/Prune(maxSlot)` methods, and when Put is called it can update the minimum value to later be able to prune from it to the maxSlot.

its probably 20 lines of code to avoid importing `gods` which is probably thousands.

Reply:
@moshe-blox treemap is used because there's a loop removing min (stale) slots that are older than `maxSlotsInState`. I think it almost always removes only one slot, but a slot gap is theoretically possible, so keeping slots sorted would help here and just one minimum value is not enough. 

The `gods` library is quite well known and takes just 1.0M, while the total size of the repo and all dependencies is 1.6G.

But I don't mind getting rid of it for simplicity and even having a loop over all values because `maxSlotsInState` is just `34` now and `pruneOldSlots` is not called often

Reply:
we decided to consider using an array instead of treemap